dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBNoSQL/AzureCosmosDBNoSQLMemoryStore.cs

```diff
@@ -0,0 +1,438 @@
+    /// <param name="vectorEmbeddingPolicy">The <see cref="VectorEmbeddingPolicy" /> to use if a collection is created. NOTE that embeddings will be stored in a property named 'embedding'.</param>
+    /// <param name="indexingPolicy">The <see cref="IndexingPolicy"/> to use if a collection is created. NOTE that embeddings will be stored in a property named 'embedding'.</param>
+    /// <param name="applicationName">The application name to use in requests.</param>
+    public AzureCosmosDBNoSQLMemoryStore(
+        CosmosClient cosmosClient,
+        string databaseName,
+        VectorEmbeddingPolicy vectorEmbeddingPolicy,
+        IndexingPolicy indexingPolicy,
+        string? applicationName = null)
+    {
+        this._cosmosClient = cosmosClient;
+        this._cosmosClient.ClientOptions.ApplicationName = applicationName;
+        this._databaseName = databaseName;
+        this._vectorEmbeddingPolicy = vectorEmbeddingPolicy;
+        this._indexingPolicy = indexingPolicy;
+    }
+
+    /// <inheritdoc/>
+    public async Task CreateCollectionAsync(
+        string collectionName,
+        CancellationToken cancellationToken = default)
+    {
+        var databaseResponse = await this._cosmosClient.CreateDatabaseIfNotExistsAsync(
+            this._databaseName, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+        var containerProperties = new ContainerProperties(collectionName, "/key")
+        {
+            VectorEmbeddingPolicy = this._vectorEmbeddingPolicy,
+            IndexingPolicy = this._indexingPolicy,
+        };
+        var containerResponse = await databaseResponse.Database.CreateContainerIfNotExistsAsync(
+            containerProperties,
+            cancellationToken: cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<string> GetCollectionsAsync(
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        using var feedIterator = this.
+            _cosmosClient
+            .GetDatabase(this._databaseName)
+            .GetContainerQueryIterator<string>("SELECT VALUE(c.id) FROM c");
+
+        while (feedIterator.HasMoreResults)
+        {
+            var next = await feedIterator.ReadNextAsync(cancellationToken).ConfigureAwait(false);
+            foreach (var containerName in next.Resource)
+            {
+                yield return containerName;
+            }
+        }
+    }
+
+    /// <inheritdoc/>
+    public async Task<bool> DoesCollectionExistAsync(
+        string collectionName,
+        CancellationToken cancellationToken = default)
+    {
+        await foreach (
+            var existingCollectionName in this.GetCollectionsAsync(cancellationToken)
+                .ConfigureAwait(false)
+        )
+        {
+            if (existingCollectionName == collectionName)
+            {
+                return true;
+            }
+        }
```

Comment:
I see in GetCollectionsAsync that there's a query being specified. Would this be able to specify a more specific query just for the name and avoid then having to bring down and iterate through the collection names on the client?

Reply:
Good catch - done.